### PR TITLE
adr: mvp light client

### PR DESCRIPTION
**Pull request by liamsi**
## Description

Rendered: https://github.com/lazyledger/lazyledger-core/blob/ismail/adr-mvp-rpc-light-client/docs/lazy-adr/adr-004-mvp-light-client.md#adr-004-data-availability-sampling-light-client

Part of #307

### CODE REVIEW

**Comment by liamsi** [comment]
Converted back to draft as I want to be a bit more precise on the IPFS integration for the DAS. Please start reviewing as this should not affect the existing sections.

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:1]:
```suggestion
# LAZY ADR 004: Data Availability Sampling Light Client
```

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:15]:
```suggestion
This ADR describes the changes necessary to augment the existing Tendermint light client implementation with DAS from a UX as well as from a protocol perspective.
```

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:13]:
```suggestion
A great introduction on the Tendermint light client (and BFT light clients in general) can be found in this series of [blog posts](https://medium.com/tendermint/everything-you-need-to-know-about-the-tendermint-light-client-f80d03856f98) as well as this [paper](https://arxiv.org/abs/2010.07031).
```

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:24]: 1. Replace dash with em dash.
2. What is "or the light clients of a **Data Availability**"???

```suggestion
For some use-cases—like DAS light validator nodes, or the light clients of a Data Availability that are run by full nodes of an Optimistic Rollup—it would even make sense that the light client (passively) participates in the consensus protocol to some extent; i.e. runs a subset of the consensus reactor to Consensus messages ([Votes](https://github.com/tendermint/tendermint/blob/bc643b19c48495077e0394d3e21e1d2a52c99548/types/vote.go#L48-L59) etc.) come in as early as possible.
```

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:27]: Link to DAHeader here instead of later, and define it in its full form.

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:48]:
```suggestion
2. request a parametrizable number of random samples.
```

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:50]:
```suggestion
If the all sampling requests succeed, the whole block is available ([with some high enough probability](https://arxiv.org/abs/1809.09044)).
```

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:66]:
```suggestion
+	daSampling     bool
```

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:110]:
```suggestion
Otherwise, we introduce yet another DB instance; something we want to avoid, especially on the long run (see [#283](https://github.com/lazyledger/lazyledger-core/issues/283)).
```

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:119]: Generally as much anonymization as possible is preferred. We'd have to think carefully of the ramifications of even revealing that a node is doing DAS vs node.

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (adlerjohn) [inline docs/lazy-adr/adr-004-mvp-light-client.md:188]: These two lines don't seem to actually be changed? Can they be removed from the diff?

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md:24]: 😅I think this is supposed to be "of a Data availability layer". 

Thanks for all the feedback 🙏

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md:1]: I'd actually prefer to remove the LAZY prefix even from previous ADRs. It doesn't add any value. It made sense when we kept the other (tendermint) ADRs around as well but not anymore. I'll remove it in the other existing adrs too.

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md:13]: Both linked resoureces do focus on BFT light clients but they introduce the general concept of other light clients (like bitcoin's spv nodes).

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md:119]: I agree. That's why I mentioned here additionally to in the "Negatives" section below.

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md]: https://github.com/lazyledger/lazyledger-core/pull/311/commits/eb64c8e0ae7c5867e92b4f899af13993175c4e0b

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md]: https://github.com/lazyledger/lazyledger-core/pull/311/commits/eca6147148fd32e25b25387a5e0b90be9efe2568

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md]: 2dddade

(a lot of `d`s this time)

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md]: a275f1e

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md]: f3006e2 ?

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (Wondertan) [inline docs/lazy-adr/adr-004-mvp-light-client.md:183]:
```suggestion
Implementations of the interface need to additionally retrieve the `DataAvailabilityHeader` for the [modified LightBlock](#lightblock).
```

Comment (Wondertan) [reply docs/lazy-adr/adr-004-mvp-light-client.md:183]: Or they can get it from `ResultCommit.SignedHeader.Commit.BlockID.DAHeader` if #304 lands before MVP, not to additional request overhead

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (Wondertan) [inline docs/lazy-adr/adr-004-mvp-light-client.md:214]:
```suggestion
Alternatively, with the exact same result, we could embed the existing `Provider` into a new interface: e.g. `DASProvider` that adds this method.
```

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (Wondertan) [inline docs/lazy-adr/adr-004-mvp-light-client.md:188]: Why not make DA one and only default case for LL? The argument that it is clear and explicit works only if there is a need to keep both ways to get LightBlock, but why do we need both ways at the same time?

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (Wondertan) [inline docs/lazy-adr/adr-004-mvp-light-client.md:217]: Why DAS light client is not a regular light client if LL is based on DAS in the first place?

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md:120]: > Things would have been so much easier if we just accessed block's data by Header.DataHash in the first place.

Well, we can still do that.

> We then would not need to store DAHeader in both Full and Light node stores, but only Header, as planned.

Wouldn't it still make sense to store the DAHeader. Especially for nodes that do not have the full block data?

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md:183]: > if #304 lands before MVP,

You mean #312?

Yeah, I intentionally didn't want to rely on #312. I can't think of a use-case where the DAHeader would be requested in it's own but I guess it certainly won't harm if it can be requested. We can remove the rpc endpoint / the additional method in the provider later if we wanted.

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md:107]: > It will be included in other data types soon.

None of the existing RPC responses a tendermint light client requests will contain the DAHeader though (the BlockID in #312 is only temporary.

> In the end, we will rely mostly on Header.DataHash instead.

True! 

> Redundantly requesting it separately is obviously not a good idea from long-living subscriptions POV, though subscribing for the whole EventNewBlock is also a bloated approach in some use-cases and EventNewBlockHeader does not contain DAHeader.

The light client itself doesn't use the event system. Are you talking about the future where the light clients will subscribe to headers as you described in https://github.com/lazyledger/lazyledger-core/issues/86#issuecomment-831182564 ?

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md:188]: > but why do we need both ways at the same time?

Some light clients will still want to able to run a tendermint light client as before. They do not need the DAHeader.

Good point about the default though.

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (marbar3778) [inline docs/lazy-adr/adr-004-mvp-light-client.md:48]: are there any assumptions that change if a user sets a lower numer?

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (Wondertan) [reply docs/lazy-adr/adr-004-mvp-light-client.md:107]: > Are you talking about the future where the light clients will subscribe to headers as you described in #86 (comment) ?

Nope. I was talking about EventBus within ll-core. The light client can subscribe to new things through RPC with EventBus as a backbone. 

From my understanding, EventNewBlock and EventNewBlockHeader are the most used RPC endpoint/subscription for the light clients to be always on track with the network, and doing separate requests for DAHeader to perform DAS on every subscription message does not make sense for me.

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (Wondertan) [reply docs/lazy-adr/adr-004-mvp-light-client.md:120]: > Well, we can still do that. It's a bit orthogonal to this ADR I guess?

Yep, that just a quick look retrospectively at how things could be done better. Nothing against this PR

Comment (Wondertan) [reply docs/lazy-adr/adr-004-mvp-light-client.md:120]: > Wouldn't it still make sense to store the DAHeader. Especially for nodes that do not have the full block data? 

If you have DataHash you can access full DAHeader, thus you can do anything you do with DAHeader, thus storing/treating it specifically is not needed, as it would be stored somewhere on the network.

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (Wondertan) [reply docs/lazy-adr/adr-004-mvp-light-client.md]: > You mean #312?
Yes

Comment (Wondertan) [reply docs/lazy-adr/adr-004-mvp-light-client.md]: > I can't think of a use-case where the DAHeader would be requested in it's own but I guess it certainly won't harm if it can be requested. We can remove the rpc endpoint / the additional method in the provider later if we wanted. It does not introduce any technical debt.

Ok

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (Wondertan) [reply docs/lazy-adr/adr-004-mvp-light-client.md:107]: @liamsi, so you didn't rely on DAHeader in BlockID as temporary, but you did rely on DAHeader instead of DataHash because it weeks away? 🤔  Kinda controversial 😄, but I understand that it was not clear before that DataHash can be dropped in instead of DAHeader so I am ok with that.

Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md:107]: > @liamsi, so you didn't rely on DAHeader in BlockID as temporary, but you did rely on DAHeader instead of DataHash because it weeks away? 🤔 Kinda controversial 😄, 

What exactly is controversial about it and what are you suggesting to do about it? Please, always keep include constructive suggestions when you criticize something. Otherwise, it is not clear what improvements could be made here.

Just to clarify: Everything in this ADR is written in a way that a) does not depend on any feature that is not clearly defined or implemented yet and b) does not introduce any technical debt. 

I don't really see why it is harmful but we can (very) easily remove the additional rpc end-point in the near future if we wanted to.

It is clear that the light client could simply do with the data hash but: 1) how to do this via IPFS is not defined nor implemented yet and 2) we wouldn't have been able to ship a DAS light client as quickly if we waited on that.

> The light client can subscribe to new things through RPC with EventBus as a backbone.

That is a valuable and cool observation but it's still not how the current default light client operates: basically the rpc endpoint of the light client "drives" the light client (it gets updated on demand if you e.g. request anything with a newer height from the light client's RPC).

That said, for future light client implementations, we could use the event bus. If you want, you can write an ADR for that (it's not urgent but we should write an ADR for different variants for light clients we want, another one is #86 for instance).

Does this make sense?

File: docs/lazy-adr/adr-004-mvp-light-client.md
Comment (liamsi) [reply docs/lazy-adr/adr-004-mvp-light-client.md:120]: > as it would be stored somewhere on the network.

That still means nodes will store the daheader in their local data stores. I guess the question here is: should we additionally store the daheader separately or not? This rather plays into adr 001 and #182. I would argue that most nodes will need the get the DAHeader and it could be useful to provide a shortcut via storing it as one blob instead of in ipld only (even if that is redundant). But maybe that is not necessary. 

We will see after we (you?) have written the adr for "datahash -> DaHeader" and after we have implemented this feature. Then we can measure latencies and decide.

**Comment by liamsi** [comment]
@Wondertan from your point of view: is there anything that should be changed before merging this? Note that we do yet have a well-established ADR process, i.e. we did not codify yet when something is really decided. So we can also amend things to the ADRs after merging. I think after the first complete implementation, we should mark the ADR as implemented and bigger changes should then require future ADRs instead.

**Review by Wondertan — APPROVED** [review]
@liamsi, sorry if this PR was waiting for my approval. I should have approved it earlier, but I was not thinking that our discussions in my commenting review are blocking, as it's not a change-requesting review(red).

**Comment by liamsi** [comment]
No worries, no need to rush with merging the ADR :-) Thanks for all the valuable feedback and discussions!

**Pull request merged**
